packages/open-next/src/adapters/cacheIntercept.ts

```diff
@@ -0,0 +1,195 @@
+    }
+
+    return _path;
+  }
+
+  matcher = (url: string): MatchedRoute | null => {
+    const route = this.prerenderedRoutes.find((r) => r.route === url);
+    if (route) {
+      return route;
+    }
+    const foundCompiled = this.compiledDynamicRoutes.find((r) => {
+      let match = r.routeRegex.exec(url);
+      if (match) {
+        return true;
+      }
+      match = r.dataRouteRegex.exec(url);
+      if (match) {
+        return true;
+      }
+      return false;
+    })?.srcRoute;
+    if (foundCompiled) {
+      const key = this.parsePath(url);
+      return {
+        key,
+        route: url,
+        revalidate: this.revalidates.get(foundCompiled) ?? 0,
+      };
+    }
+    return null;
+  };
+
+  async revalidate(uri: string, host?: string) {
+    await new Promise<void>((resolve, reject) => {
+      httpsRequest(`https://${host}${uri}`, {
+        method: "HEAD",
+        headers: { "x-prerender-revalidate": this.preview },
+      })
+        .on("error", (err) => reject(err))
+        .end(() => resolve());
+    });
+  }
+
+  async handler(event: ProxyEvent): Promise<ProxyResult | false> {
+    const startTime = Date.now();
+    // const request = event.Records[0].cf.request;
+    const uri = event.rawPath;
+    if (event.headers["x-prerender-revalidate"] === this.preview) {
+      return false;
+    }
+
+    console.log("Matched uri:", uri);
+
+    const matchedRoute = this.matcher(uri);
+    const method = event.method;
+    if (method !== "GET" && method !== "HEAD") {
+      return false;
+    }
+    if (matchedRoute) {
+      const { revalidate, key } = matchedRoute;
+      const isDataRoute = uri.startsWith("/_next/data");
+
+      try {
+        const bucketName = process.env.CACHE_BUCKET_NAME;
+        const { Body, LastModified } = await this.s3.send(
+          new GetObjectCommand({
+            Bucket: bucketName,
+            Key: `${this.buildId}${key}${isDataRoute ? ".json" : ".html"}`,
+          })
+        );
+        if (!Body) throw new Error("No body found in s3 response.");
+        if (!LastModified) throw new Error("No LastModified found in s3 response.");
+
+        const body = await Body.transformToString();
+        let isStale = false;
+        const expirationDate = LastModified.getTime() + revalidate * 1000;
+        const now = Date.now();
+        const remaining = ((expirationDate - now) / 1000).toFixed(0);
+
+        if (expirationDate < now) {
+          // Stale, revalidate
+          await this.revalidate(uri, event.headers.host);
+          isStale = true;
+        }
+
+        console.log(`Serving ${uri} from s3 cache in ${Date.now() - startTime}ms`);
+        return {
+          statusCode: 200,
+          body: method === "GET" ? body : "",
+          headers: {
+            "Content-Type": isDataRoute ? "application/json" : "text/html",
+            "Cache-Control": isStale
+              ? "s-maxage=0"
+              : `s-max-age=${remaining}, stale-while-revalidate`,
```

Comment:
Looks like a typo, "s-max-age" instead of "s-maxage"

```suggestion
              : `s-maxage=${remaining}, stale-while-revalidate`,
```